Does not build on Mac OS X

Hi All,

Has anyone build `libzim` on a Mac? Is it a supported platform for dev?

I wanted to test some things out (exploring `pybind11`) and after installing the dependencies and running `ninja` I ran into this error:

```
...
FAILED: src/zim@sha/buffer.cpp.o
c++  -Isrc/zim@sha -Isrc -I../src -Iinclude -I../include -I/usr/local/Cellar/xz/5.2.3/include -I/usr/local/Cellar/xapian/1.4.4/include -I/usr/local/Cellar/icu4c/59.1/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++11 -O0 -g -MMD -MQ 'src/zim@sha/buffer.cpp.o' -MF 'src/zim@sha/buffer.cpp.o.d' -o 'src/zim@sha/buffer.cpp.o' -c ../src/buffer.cpp
../src/buffer.cpp:45:68: error: use of undeclared identifier 'MAP_POPULATE'
```

It seems like `MAP_POPULATE` is a linux-only thing: https://stackoverflow.com/questions/35568112/use-of-undeclared-identifier-map-populate but I know libzim or did I misconfigure something?

I'm fine working under linux too, just wanted to check if someone is doing dev on `libzim` on a Mac and knows a workaround.

Here are the dependency-installation steps through homebrew:
```
brew install zlib
brew install xapian
brew upgrade icu4c
meson . build
```
I'm on Yosemite 10.10.5

Thanks for the pointer to `kiwix-build`. Very nice build system!

Instructions seem up to date. I ran `./kiwix-build.py --target-platform native_dyn` and pretty much everything worked except `libzim` compilation with two errors:
```
In file included from ../../SOURCE/libzim/include/zim/file_reader.h:23:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:4032:35: error: no viable overloaded '='
                __e->__weak_this_ = *this;
                ~~~~~~~~~~~~~~~~~ ^ ~~~~~
```
and
```
[12/33] c++  -Isrc/zim@sha -Isrc -I../../SOURCE/libzim/src -Iinclude -I../../SOURCE/libzim/include -I/Users/ivan/Projects/FLECode/wikimania-hackathon/kiwix-build/BUILD_native_dyn/INSTALL/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++11 -O0 -g -MMD -MQ 'src/zim@sha/buffer.cpp.o' -MF 'src/zim@sha/buffer.cpp.o.d' -o 'src/zim@sha/buffer.cpp.o' -c ../../SOURCE/libzim/src/buffer.cpp
FAILED: src/zim@sha/buffer.cpp.o
c++  -Isrc/zim@sha -Isrc -I../../SOURCE/libzim/src -Iinclude -I../../SOURCE/libzim/include -I/Users/ivan/Projects/FLECode/wikimania-hackathon/kiwix-build/BUILD_native_dyn/INSTALL/include -Xclang -fcolor-diagnostics -pipe -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++11 -O0 -g -MMD -MQ 'src/zim@sha/buffer.cpp.o' -MF 'src/zim@sha/buffer.cpp.o.d' -o 'src/zim@sha/buffer.cpp.o' -c ../../SOURCE/libzim/src/buffer.cpp
../../SOURCE/libzim/src/buffer.cpp:45:68: error: use of undeclared identifier 'MAP_POPULATE'
  _data = (char*)mmap(NULL, size + _offset, PROT_READ, MAP_PRIVATE|MAP_POPULATE, fd, pa_offset); 
```
full log: https://gist.github.com/ivanistheone/2d1db6c4618d702842f13d371b9ef7da


I understand Mac OS X is currently not an official target so don't waste time debugging this, this was just me playing around. Will try with a more recent gcc this afternoon.

--- a/src/buffer.cpp
+++ b/src/buffer.cpp
@@ -42,7 +42,13 @@ MMapBuffer::MMapBuffer(int fd, std::size_t offset, std::size_t size):
 {
   std::size_t pa_offset = offset & ~(sysconf(_SC_PAGE_SIZE) - 1);
   _offset = offset-pa_offset;
-  _data = (char*)mmap(NULL, size + _offset, PROT_READ, MAP_PRIVATE|MAP_POPULATE, fd, pa_offset);
+#if defined(__APPLE__)
+  #define MAP_FLAGS MAP_PRIVATE
+#else
+  #define MAP_FLAGS MAP_PRIVATE|MAP_POPULATE
+#endif
+  _data = (char*)mmap(NULL, size + _offset, PROT_READ, MAP_FLAGS, fd, pa_offset);
+#undef MAP_FLAGS
 }
 
 MMapBuffer::~MMapBuffer()

